### PR TITLE
KEYCLOAK-17637 Client Scope Policy for authorization service

### PR DESCRIPTION
**Pull request by y-tabata**
JIRA: https://issues.redhat.com/browse/KEYCLOAK-17637

# Spec
- Evaluate based on client scopes included as "scope" claim in the access token.
- Like the role policy, there is a `required` check box.

### CODE REVIEW

**Comment by y-tabata** [comment]
Docs PR: https://github.com/keycloak/keycloak-documentation/pull/1148

File: authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/clientscope/ClientScopePolicyProviderFactory.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            if (event instanceof ClientScopeRemovedEvent) {
+                KeycloakSession keycloakSession = ((ClientScopeRemovedEvent) event).getKeycloakSession();
+                AuthorizationProvider provider = keycloakSession.getProvider(AuthorizationProvider.class);
+                StoreFactory storeFactory = provider.getStoreFactory();
+                PolicyStore policyStore = storeFactory.getPolicyStore();
+                ClientScopeModel removedClientScope = ((ClientScopeRemovedEvent) event).getClientScope();
+                ResourceServerStore resourceServerStore = storeFactory.getResourceServerStore();
+                RealmModel realm = removedClientScope.getRealm();
+
+                realm.getClientsStream().forEach(clientModel -> {
```
Comment (pedroigor) [inline authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/clientscope/ClientScopePolicyProviderFactory.java:76]: We don' need to iterate over clients but over policies with the expected type. Could you please, look at this branch https://github.com/pedroigor/keycloak/tree/KEYCLOAK-17637?

Pretty much the same logic but iterating directly over client scope policies, regardless of the client.

**Review by pedroigor — CHANGES_REQUESTED** [review]
@y-tabata Thanks for this one.

Please, take a look at https://github.com/pedroigor/keycloak/tree/KEYCLOAK-17637 for some changes to how policies are updated/removed when scopes are deleted. As well as one more test to cover this scenario.

File: authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/clientscope/ClientScopePolicyProviderFactory.java (hunk 1 above)
Comment (stianst) [reply authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/clientscope/ClientScopePolicyProviderFactory.java:76]: +1 realm.getClientsStream will be banned going forward as it just doesn't scale with many clients ;)

**Comment by y-tabata** [comment]
@pedroigor @stianst Thanks for your review. I changed the code according to https://github.com/pedroigor/keycloak/tree/KEYCLOAK-17637.

**Review by pedroigor — DISMISSED** [review]
@y-tabata Thanks.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/ClientScopePolicyTest.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    @Before
+    public void configureAuthorization() throws Exception {
+        createResource("Resource A");
+        createResource("Resource B");
+
+        createClientScopePolicy("Client Scope foo Policy", "foo", "bar");
+        createClientScopePolicyAndLastOneRequired("Client Scope bar Policy", "foo", "bar");
+        createClientScopePolicy("Client Scope To Remove Policy", "to-remove-a", "to-remove-b");
```
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/ClientScopePolicyTest.java:78]: The test failure is cause by this line.

Could you please:

* Move `createClientScopePolicy("Client Scope To Remove Policy", "to-remove-a", "to-remove-b");` to the `testRemovePolicyWhenRemovingScope`?

Comment (y-tabata) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/ClientScopePolicyTest.java]: @pedroigor 
Changed.

**Comment by y-tabata** [comment]
@pedroigor Thanks for your review.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@y-tabata Thank you.

**Pull request merged**